### PR TITLE
Upload migration payload by chunks

### PR DESCRIPTION
**Pull request by nhan-nguyen-se**
- Upload starting payload right from the begin of migration. This payload contains server app version and source app information.
- Use CQL search API to fetch pages, 1000 pages each chunk. For each chunk, upload the data payload which contains page IDs
- When all data payloads has been uploaded, then upload an ending payload which contain total non-empty chunks count. Cloud side can use this information to know if all data payloads has been uploaded and start to process them.

Payload types (starting, ending, data) are distinguished using migration label.

Cloud side PR: https://github.com/Atlas-Authority/MarkdownMacroCloud/pull/204

### CODE REVIEW

**Comment by boris-moduscreate** [comment]
@sdegutis make sure you're aware of this and associated Cloud changes as you're looking into how to handle parallel migration in https://github.com/Atlas-Authority/MarkdownMacroCloud/issues/202

**Comment by sdegutis** [comment]
👍

File: src/main/java/com/atlassian/plugins/confluence/markdown/ccma/CqlSearchIterator.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    private PageRequest nextPage(PageResponse<Content> pageResponse) {
+        final int nextStart = Optional.ofNullable(pageResponse)
+                .map(PageResponse::getPageRequest)
+                .map(page -> page.getStart() + page.getLimit())
+                .orElse(0);
+        return new SimplePageRequest(nextStart, pageSize);
+    }
+
+    private boolean hasMore(PageResponse<Content> pageResponse) {
+        return Optional.ofNullable(pageResponse).map(PageResponse::hasMore).orElse(true);
```
Comment (khanguyen88) [inline src/main/java/com/atlassian/plugins/confluence/markdown/ccma/CqlSearchIterator.java:66]: If ``pageResponse`` is null, does it mean we have no more data? If so ``orElse(false)`` is the correct value.

File: src/main/java/com/atlassian/plugins/confluence/markdown/ccma/PageIterator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    private PageRequest nextPage(PageResponse<T> pageResponse) {
+        final int nextStart = Optional.ofNullable(pageResponse)
+                .map(PageResponse::getPageRequest)
+                .map(page -> page.getStart() + page.getLimit())
+                .orElse(0);
+        return new SimplePageRequest(nextStart, pageSize);
+    }
+
+    private boolean hasMore(PageResponse<T> pageResponse) {
+        return Optional.ofNullable(pageResponse).map(PageResponse::hasMore).orElse(true);
```
Comment (khanguyen88) [inline src/main/java/com/atlassian/plugins/confluence/markdown/ccma/PageIterator.java:63]: same rational with ``orElse`` here

File: src/main/java/com/atlassian/plugins/confluence/markdown/ccma/CqlSearchIterator.java (hunk 1 above)
Comment (nhan-nguyen-se) [reply src/main/java/com/atlassian/plugins/confluence/markdown/ccma/CqlSearchIterator.java]: This class is redundant. Removed.

File: src/main/java/com/atlassian/plugins/confluence/markdown/ccma/PageIterator.java (hunk 2 above)
Comment (nhan-nguyen-se) [reply src/main/java/com/atlassian/plugins/confluence/markdown/ccma/PageIterator.java:63]: PageResponse is only null in first iteration, so should be `true` in this case.

**Review by khanguyen88 — APPROVED** [review]

**Review by sdegutis — APPROVED** [review]

**Comment by boris-moduscreate** [comment]
This should not be released until the Cloud changes are also live right?

**Comment by nhan-nguyen-se** [comment]
Server and cloud must be released together. I have tested on staging and it work fine (migrate from my local server to staging cloud app). I'm working on replicate the change to iframe-for-confluence but got issue with dependency injection on that repo. Working with @khanguyen88 to resolve it.

**Pull request merged**